### DISCUSSION TITLE
Translation and sync automation

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#29949] [Announcements]
This is a raw idea and RFC discussion.

With an inlined translation format, we can automatically sync languages with `en` without the need for manual maintenance and ensure we have a markdown-native flow while being possible to integrate with cloud-based systems and docus for augmented improvements like showing translate button for each section!

## Example

[content/ja/intro.md]

```md
---
description.en: 'Build your next Vue.js application with confidence using Nuxt. An open-source framework making web development simple and powerful.'
description.ja: 'Nuxt を使って信頼ある次世代の Vue.js アプリケーションを構築する Web 開発をシンプルにかつパワフルにするオープンソースフレームワーク'

---

<!-- [translated] # Intro to Nuxt -->
# Nuxtの紹介

<!-- [translated] ## Section 1 -->
## セクション1

<!-- [translated] This is paragraph 1 -->
これはパラグラフ1です

<!-- [attention] ## Section 2 (optional) -->
## セクション2

<!-- [translated] This is paragraph 2 -->
これはパラグラフ2です

<!--
[translateme]
::alert{type=warning}
Hello
::
-->
::alert{type=warning}
Hello
::
```

## Magic translation comments

Before each title and before its contents, the automated script adds a magic comment which is basically a snapshot of the latest reference language (`en`) and a tag that specifies its status. This helps both automation and maintainers to quickly read, compare and do translations.

### Translation tags

There are 3 possible status tags determined during sync process by comparing latest reference to new one.
- `[translateme]`: Content is either new or could not match or by fuzzy search has difference above the threshold 
- `[attention]`:  Difference between last reference and next sync, has some difference but below the threshold to be replaced with en
- `[translated]`: Last and next sync reference was same. Therefore if status was translated, it is still translated without need of attention

## Sync algorithm

- Fetching reference (`en`) content from all sources
- Parse each file by section
- Compare reference comments in languages with original and detect status (`translateme`, `attention` or `translated`)
- Update reference comments and actual content to `en` if the status is `translateme`
- Commit changes

## Markdown meta

The current idea is to support an object with each language.

**Comment by Geminii** [comment]
Thanks for this nice purpose about translations diff 🙏

Does we really need `[translated]` tag if the translation is already the same ? It can be confusing and not a real interest to precise it in this case. What do you think about it ? 

For more detail, when the `sync algorithm` will be executed ? On each push on specific branch like `main` ?

**Reply by Intevel** [reply]
You can check out the platform Crowdin, maybe that's an option too.

**Reply by Geminii** [reply]
We have already used in the past with @Atinux and it wasn't a good solution 😅

**Reply by pi0** [reply]
This is a nice idea @Geminii to omit tag when translated. I mainly wanted to keep it so we can detect translation comments and distinguish them from other comments. 

@Intevel Indeed we were using several platforms in past. The main issue is when depending on a cloud-based solution compared to GitHub, we are bound to platform possibilities while with a markdown-only approach we never have this risk yet will be able to integrate with cloud based solutions (think of git merge conflict annotations and how IDEs support them)

**Comment by kazupon** [comment]
I think it's a really good idea!!  ❤️ 

This idea is based on docs syntax and requires Magic translation comments.
Therefore, it may look a bit complicated because docs itself, docs syntax, and Magic translation comments are in makrdown.
This may seem a bit complicated, especially for those who have translated using the simple makrdown.
It may be better to use a different syntax for Magic translation comments.

**Reply by pi0** [reply]
Indeed we bring easier and consistent to the translation process with the cost of adding more verbosity. Without inlined reference we have two issues:
- A mapping database (external file) is needed to keep track of the status or we should let maintainers do it with a classic manual process 
- The translators have to check sections one by one finding of what is missing
(while with inlined reference, both automated systems and translators can quickly check what was original sentence)

What do you think about more semantic annotation to make it more readable for new users?

```md
<!-- translate.en: # Header 1 -->
```

**Reply by AdamModus** [reply]
Maybe @kazupon has experience on this already, so please let me know if you got ideas/suggestions!

If we were to build a tool (does not have to necessarily be related to Nuxt itself) that is more geared towards managing markdown based documentation, each "translatable" single section could have 3 base states:

- not translated
- translated: for this, there could be sub states to translated to language X, language Y, etc...)
- outdated: it was translated before but thr original section has changed.

This could help keeping an overview of the documentation and its current state. An outdated section could then be flagged for any maintainers to try to either update the translation / get help to do it.

The issue with this is that keeping track of all these states in the MD file itself could be tricky (too much visual pollution and makes it difficult to read).

Your comment regarding an external database file @pi0, made me think of this. Was this something you already had in mind?

**Reply by AdamModus** [reply]
Maybe a way to match if a translated section is up to date or not is to generate a hash based on its content (assuming the base is always English).

The moment said content changes, all translations that point to it but have different "original content hashes". But that also means the each section needs to connect to its translations via some sort of key (order or section within original MD file might be enough)

**Reply by pi0** [reply]
Using hash-only, we cannot do fuzzy checks anymore (for the threshold of acceptable change)

We can have status + content ref like `<-- [translated] e123:en/intro.md:10:20 -->` instead of actually inlining contents. This can also be a configurable behavior.

What I'm afraid of about implicit refs, is that this would make the process more likely to human errors because we are still depending on **manual and optional** review. A maintainer can simply mark it as `[translated]` without bothering to check updates while with inlining contents, they can clearly see it above and if translation is not good enough, next person can use it as a reference without bothering to check another file.

**Reply by AdamModus** [reply]
~~I guess a relevant question is also how granular would you would like each translatable bit to be, i.e. each paragraph is a separate bit, file per file basis, everything under a sub-header is a bit, etc.~~

~~And should all these bits be checked separately.~~

Ignore that. You had already answered that initially 😅

**Reply by kazupon** [reply]
> What do you think about more semantic annotation to make it more readable for new users?

I agree with semantic annotation. With its introduction, new users will be able to read markdown content. In addition to that, it also allows a synchronization mechanism based on annotations with the script.

However, we need to define the syntax of semantic annotation so that the readability of the content is not reduced. One of the idea would be to use comment annotation like jsdoc.

### embed content annotation syntax
short-content:
```
<!--
  en: '# Header 1'
  status: translateme
-->
```

long-content:
```
<!--
  en: `
    ::alert{type=warning}
    Hello
    ::
  `
  status: translateme
-->
```

### reference content annotation syntax
```
<!--
  ref: en/intro.md
  line: 10
  column: 20
  status: translated
-->
# ヘッダー１
```

So, For we implement semantic annotation, we need to provide syntax highlight support for an IDE such as VSCode.
We already have a VSCode extension for docus, we just need to extend it! :)

**Comment by AdamModus** [comment]
I like the idea as well. It is a nice in-between a key-translation pair that we see with typical i18n-related solutions that work very well for smaller scopes of translation but become harder when you have to translate multiple paragraphs in one go.

question/hypothesis:

- would you prefer this to happen run-time? e.g. i'm a dev and i run `**localhost:4000/nuxt-docs**` which gives me english website by default and switching to `**localhost:4000/fr/nuxt-docs**` calls this translation on-the-go and servers everything.

- would you rather have a "generate translated docs" script that would go through all the docs, sieve through all the existing translations and generate everything that i can. said script could be called on github actions on locally for contributors to test their translations.

**Reply by pi0** [reply]
> would you prefer this to happen run-time? e.g. i'm a dev and i run **localhost:4000/nuxt-docs** which gives me english website by default and switching to **localhost:4000/fr/nuxt-docs** calls this translation on-the-go and servers everything.

With either multi source support from docus or temporary an internal script, we can render docs in dev as production. `localhost:4000/en` and `localhost:4000/fr` both will be available while working on docs.

> would you rather have a "generate translated docs" script that would go through all the docs, sieve through all the existing translations, and generate everything that I can. said script could be called on GitHub actions locally for contributors to test their translations.

Exactly. The script is supposed to be triggered by Github actions periodically or ideally with a commit hook to immediately check differences, apply status changes and notify language maintainers. With such automation, we would never be worried about outdated translations. (Ambitiously we can potentially use google cloud translation API to create a trained model based on initial translations that can automatically apply domain-specific translations)

BTW we can also give that "manual button" to maintainers running script locally if the action goes wrong or needs a manual sync for any reason.

**Reply by AdamModus** [reply]
I like with the manual button thing. I think it would make it easier for contributors to test locally what they are doing, in a prod like situation

**Reply by kazupon** [reply]
> would you rather have a "generate translated docs" script that would go through all the docs, sieve through all the existing translations and generate everything that i can. said script could be called on github actions on locally for contributors to test their translations.

I think really so too. Recently, translation services are getting more and more accurate. 
Especially for our software documents, It's not wierd to read the translated docs

I would add to this idea, so if you have outdated translated content, we should also translate it with a translation service and add it as an annotation along with the `machine translated` status.

**Reply by AdamModus** [reply]
@kazupon, I had not thought of that! that's a really good idea!
From the translation contributions that I've done, there were moments that I couldn't think how to explain certain phrases and google translator helped me a lot, i'd then take that translation and improve it but it did "unlock" me at that moment.
